### PR TITLE
EIP-2028: calldata gas cost reduction

### PR DESCRIPTION
**Pull request by patogallaiovlabs**
## Description

Implementation of the EIP-2028 proposal. The proposal is about reducing the gas cost of Calldata (GTXDATANONZERO = TX_NO_ZERO_DATA) from its current value of 68 gas per non-zero byte to 16. This translates to an increase of the block size (measured in bytes).

## Motivation and Context

The EIP-2028 is a scalability proposal for the Ethereum network, thinking in the L2 as a way to scale the network. The commitments from L2 networks are placed in the L1 by a transaction with the collected data as a calldata. That is why we need to reanalyze and reduce the gas cost of the bytes of the calldata field.
Readers who know nothing about rollups can figure out that this proposal can reduce some operating costs for L2 rollups. It will be harder for them to infer the other impact - for any given RSK block gas limit, this reduction allows us to convey more L2 transaction data  (e.g. token transfers), which improves the transaction processing capacity (throughput) of the rollup. 

## How Has This Been Tested?

**It has not been tested.** 
One of the possible issues is the **network delay** of transmitting the block between peers, while the other is the **block processing time**. 
Tested should be oriented in those two aspects.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ X ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ X ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ X ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 1]
```diff
@@ -204,7 +205,11 @@ public long transactionCost(Constants constants, ActivationConfig.ForBlock activ
         long nonZeroes = this.nonZeroDataBytes();
         long zeroVals = ListArrayUtil.getLength(this.getData()) - nonZeroes;
 
-        return (this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION) + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * GasCost.TX_NO_ZERO_DATA;
+        return (this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION) + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * getTxNoZeroData(activations);
```
Comment (rmoreliovlabs) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:208]: Can we make this more readable?

A suggestions can be:
```
long transactionCost =  this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION;

long transactionZeroData = getTxNoZeroData(activations);

long totalTransactionCost = transactionCost + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * transactionZeroData;

return totalTransactionCost;
```

I think this just to make it easier to read the calculation process.

**Review by rmoreliovlabs — CHANGES_REQUESTED** [review]

**Review by rmoreliovlabs — CHANGES_REQUESTED** [review]
Also some tests are failing due to a bad mock for `config.getActivationConfig()` I think you should keep the feature disabled and only activate it for the new tests you will add so you can assert the new calculated values.

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 1 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:208]: Done

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 2]
```diff
@@ -204,7 +205,15 @@ public long transactionCost(Constants constants, ActivationConfig.ForBlock activ
         long nonZeroes = this.nonZeroDataBytes();
         long zeroVals = ListArrayUtil.getLength(this.getData()) - nonZeroes;
 
-        return (this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION) + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * GasCost.TX_NO_ZERO_DATA;
+        long transactionCost =  this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION;
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:208]:
```suggestion
        long transactionCost = this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION;
```

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 3]
```diff
@@ -204,7 +205,15 @@ public long transactionCost(Constants constants, ActivationConfig.ForBlock activ
...
+
+        long transactionZeroData = getTxNoZeroData(activations);
+
+        return transactionCost + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * transactionZeroData;
+    }
+
+    private static long getTxNoZeroData(ActivationConfig.ForBlock activations) {
+        return activations.isActive(ConsensusRule.RSKIPXXX)? GasCost.TX_NO_ZERO_DATA_EIP2028 : GasCost.TX_NO_ZERO_DATA;
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:216]:
```suggestion
        return activations.isActive(ConsensusRule.RSKIPXXX) ? GasCost.TX_NO_ZERO_DATA_EIP2028 : GasCost.TX_NO_ZERO_DATA;
```

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 4]
```diff
@@ -204,7 +205,15 @@ public long transactionCost(Constants constants, ActivationConfig.ForBlock activ
...
 
-        return (this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION) + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * GasCost.TX_NO_ZERO_DATA;
+        long transactionCost =  this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION;
+
+        long transactionZeroData = getTxNoZeroData(activations);
+
+        return transactionCost + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * transactionZeroData;
+    }
+
+    private static long getTxNoZeroData(ActivationConfig.ForBlock activations) {
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:215]:
```suggestion
    private static long getTxNonZeroDataCost(ActivationConfig.ForBlock activations) {
```

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java [hunk 5]
```diff
@@ -204,7 +205,15 @@ public long transactionCost(Constants constants, ActivationConfig.ForBlock activ
         long nonZeroes = this.nonZeroDataBytes();
         long zeroVals = ListArrayUtil.getLength(this.getData()) - nonZeroes;
 
-        return (this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION) + zeroVals * GasCost.TX_ZERO_DATA + nonZeroes * GasCost.TX_NO_ZERO_DATA;
+        long transactionCost =  this.isContractCreation() ? GasCost.TRANSACTION_CREATE_CONTRACT : GasCost.TRANSACTION;
+
+        long transactionZeroData = getTxNoZeroData(activations);
```
Comment (Vovchyk) [inline rskj-core/src/main/java/org/ethereum/core/Transaction.java:210]:
```suggestion
        long txNonZeroDataCost = getTxNonZeroDataCost(activations);
```

File: rskj-core/src/main/resources/config/devnet.conf [hunk 6]
```diff
@@ -12,7 +12,8 @@ blockchain.config {
         hop400 = 0,
         hop401 = 0,
         fingerroot500 = 0,
-        tbd600 = 0
+        tbd600 = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/devnet.conf:15]:
```suggestion
        tbd600 = 0
```
for devnet and regtest it should be `0`

File: rskj-core/src/main/resources/config/devnet.conf [hunk 7]
```diff
@@ -12,7 +12,8 @@ blockchain.config {
...
+        xxx = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/devnet.conf:16]: not needed

File: rskj-core/src/main/resources/config/main.conf [hunk 8]
```diff
@@ -13,6 +13,7 @@ blockchain.config {
         hop401 = 4976300,
         fingerroot500 = 5468000,
         tbd600 = -1
+        xxx = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/main.conf:16]: not needed

File: rskj-core/src/main/resources/config/regtest.conf [hunk 9]
```diff
@@ -12,7 +12,8 @@ blockchain.config {
         hop400 = 0,
         hop401 = 0,
         fingerroot500 = 0
-        tbd600 = 0
+        tbd600 = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/regtest.conf:15]:
```suggestion
        tbd600 = 0
```
should be 0 for regtest

File: rskj-core/src/main/resources/config/regtest.conf [hunk 10]
```diff
@@ -12,7 +12,8 @@ blockchain.config {
...
+        xxx = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/regtest.conf:16]: not needed

File: rskj-core/src/main/resources/config/testnet.conf [hunk 11]
```diff
@@ -13,6 +13,7 @@ blockchain.config {
         hop401 = 3362200,
         fingerroot500 = 4015800,
         tbd600 = -1
+        xxx = -1
```
Comment (Vovchyk) [inline rskj-core/src/main/resources/config/testnet.conf:16]: not needed

**Review by Vovchyk — CHANGES_REQUESTED** [review]

File: rskj-core/src/main/resources/config/testnet.conf (hunk 11 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/config/testnet.conf:16]: Removed

File: rskj-core/src/main/resources/config/regtest.conf (hunk 10 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/config/regtest.conf:16]: Removed

File: rskj-core/src/main/resources/config/main.conf (hunk 8 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/config/main.conf:16]: Removed

File: rskj-core/src/main/resources/config/devnet.conf (hunk 7 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/resources/config/devnet.conf:16]: Removed

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 4 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:215]: Done

File: rskj-core/src/main/java/org/ethereum/core/Transaction.java (hunk 2 above)
Comment (rmoreliovlabs) [reply rskj-core/src/main/java/org/ethereum/core/Transaction.java:208]: Done

**Review by Vovchyk — COMMENTED** [review]
lgtm!

**Review by asoto-iov — DISMISSED** [review]

**Review by Vovchyk — APPROVED** [review]

**Review by rmoreliovlabs — APPROVED** [review]
LGTM

**Pull request merged**